File: leetcode-solutions/ruby/278. First Bad Version.rb
# https://leetcode.com/problems/first-bad-version/

# The is_bad_version API is already defined for you.
# @param {Integer} version
# @return {boolean} whether the version is bad
# def is_bad_version(version):

# @param {Integer} n
# @return {Integer}
def first_bad_version(n)
  low = 1
  high = n
  
  while low < high
      mid = (low + high) / 2
      r = is_bad_version(mid)
      if r 
          high = mid
      else
          low = mid + 1
      end
  end
 
  low
end

File: leetcode-solutions/python/Algorithm-1/278. First Bad Version.py
# https://leetcode.com/problems/first-bad-version/

# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
# def isBadVersion(version):

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        s, e = 1, int(n)
        while s < e:
            mid = (e - s) // 2 + s
            r = isBadVersion(mid)
            if r:
                e = mid
            else:
                s = mid + 1
        # print(s)
        return s
